Mlem/Models/UserSettings.swift

```diff
@@ -0,0 +1,449 @@
+//
+//  UserSettings.swift
+//  Mlem
+//
+//  Created by Taylor Geisse on 6/16/23.
+//
+
+import Foundation
+import SwiftUI
+
+struct UserSettings {
+    let shouldShowWebsitePreviews = Setting(key: "shouldShowWebsitePreviews", defaultValue: true)
+    let shouldShowWebsiteFaviconAtAll = Setting(key: "shouldShowWebsiteFaviconAtAll", defaultValue: true)
+    let shouldShowWebsiteHost = Setting(key: "shouldShowWebsiteHost", defaultValue: true)
+    let shouldShowWebsiteFavicons = Setting(key: "shouldShowWebsiteFavicons", defaultValue: true)
+    
+    let shouldShowCompactPosts = Setting(key: "shouldShowCompactPosts", defaultValue: false)
+    
+    let shouldShowUserAvatars = Setting(key: "shouldShowUserAvatars", defaultValue: true)
+    let shouldShowCommunityIcons = Setting(key: "shouldShowCommunityIcons", defaultValue: true)
+    let shouldShowCommunityHeaders = Setting(key: "shouldShowCommunityHeaders", defaultValue: false)
+    
+    let voteComplexStyle = Setting(key: "voteComplexStyle", defaultValue: VoteComplexStyle.standard)
+    
+    let defaultCommentSorting = Setting(key: "defaultCommentSorting", defaultValue: CommentSortTypes.top)
+    
+    let hasUndergoneLegacyAccountDeletion_debug_3 = Setting(key: "hasUndergoneLegacyAccountDeletion_debug_3", defaultValue: false)
+    
+    static let shared = UserSettings()
+    private init() {}
+    
+    struct Setting<Value> {
+        let key: String
+        let defaultValue: Value
+        
+        init(key: String, defaultValue: Value) {
+            self.key = key
+            self.defaultValue = defaultValue
+        }
+        
+        init(key: String) where Value: ExpressibleByNilLiteral {
+            self.key = key
+            self.defaultValue = nil
+        }
+    }
+}
+
+
+// MARK: - AppStorage Extension for Non-Optionals
```

Comment:
This is a lot of boilerplate that seems mostly the same pattern/code, is there a way to dedup that?

Reply:
It does seem like a lot of boilerplate, but it is needed to match AppStorage's initializers: https://developer.apple.com/documentation/swiftui/appstorage

I can group extensions based on @available groupings and move where clauses, if that would be cleaner at least?

The UserDefault might be able to be reduced, but I submitted the PullRequest the way I did to match the initializers from AppStorage.

Reply:
I say we have a think about this, I like the idea - and I like the use of keypaths over string literals, but I wonder if we could achieve the same result with a wrapper?

Could we not have an object which declares our preferences, and accessible by keypath, like you've done here - but it just stores the values using `AppStorage` internally?

We wouldn't be able to declare a value that wasn't supported by `AppStorage` because the compiler would raise an error, avoiding the need to extend over every initialiser?

The only sticking point I can think of is the observability, but that should be solvable? 🤔

Reply:
There is the solution that I have used before, but it is hard to get working with enums: https://www.avanderlee.com/swift/appstorage-explained/
The only way I have gotten it to work with enums (and nils) is to write explicit cases in the get / set accessors to check for my enums while casting.

It also does not provide type conformance, which means that trying to store a value that is not supported by UserDefaults will not fail until runtime.

The @UserDefault propretyWrapper I've written is not observable, true. The @AppStorage abstracted initializers are observable, though, in SwiftUI as the string counterparts are. This is where the alternative I've linked above comes at an advantage. I believe it can be used to observe changes outside of SwiftUI via Combine.

Reply:
> Could we not have an object which declares our preferences, and accessible by keypath, like you've done here - but it just stores the values using AppStorage internally?
> 
> We wouldn't be able to declare a value that wasn't supported by AppStorage because the compiler would raise an error, avoiding the need to extend over every initialiser?

I just tried a bit and AppStorage does not expose, from what I can tell, anyway to observe changes.

Reply:
I think I found a way. Please hold while I test it.